File: apps/web/src/lib/permit.ts
import { getToken } from "@/lib/auth";
import { getApiBase } from "@/lib/config";

export type RoutePoint = {
  name?: string | null;
  address?: string | null;
  lat?: number | null;
  lon?: number | null;
};

export type CargoSpec = {
  length_m: number;
  width_m: number;
  height_m: number;
  mass_kg: number;
  axle_loads_kg: number[];
};

export type PermitAnalysis = {
  permit_required: boolean;
  podd_required: boolean;
  risk_level: string;
  risk_score: number;
  route_summary: string;
  restrictions: string[];
  special_conditions: string[];
  alternative_routes: string[];
  legal_references: string[];
  dispatcher_actions: string[];
  confidence: number;
};

export type PermitRequest = {
  id: string;
  status: string;
  origin: RoutePoint;
  destination: RoutePoint;
  cargo: CargoSpec;
  analysis: PermitAnalysis | null;
  risk_score: number | null;
  external_permit_number: string | null;
  dispatcher_notes: string | null;
  created_at: string;
};

function authHeaders(): HeadersInit {
  const token = getToken();
  if (!token) throw new Error("Не авторизован — войдите снова");
  return {
    Authorization: `Bearer ${token}`,
    "Content-Type": "application/json",
  };
}

async function readError(response: Response, fallback: string): Promise<string> {
  const body = await response.json().catch(() => ({}));
  if (typeof body.detail === "string") return body.detail;
  if (Array.isArray(body.detail)) {
    return body.detail.map((d: { msg?: string }) => d.msg ?? String(d)).join(", ");
  }
  return fallback;
}

export async function analyzeRoute(payload: {
  origin: RoutePoint;
  destination: RoutePoint;
  cargo: CargoSpec;
  persist?: boolean;
}): Promise<{ analysis: PermitAnalysis; permit_request_id: string | null }> {
  const response = await fetch(`${getApiBase()}/api/v1/agents/permit/analyze-route`, {
    method: "POST",
    headers: authHeaders(),
    body: JSON.stringify({ ...payload, persist: payload.persist ?? true }),
  });
  if (!response.ok) throw new Error(await readError(response, "Ошибка анализа маршрута"));
  return response.json();
}

export async function listPermitRequests(): Promise<{ items: PermitRequest[]; total: number }> {
  const response = await fetch(`${getApiBase()}/api/v1/agents/permit/requests`, {
    headers: authHeaders(),
  });
  if (!response.ok) throw new Error(await readError(response, "Не удалось загрузить заявки"));
  return response.json();
}

export async function approvePermitRequest(
  id: string,
  payload: { notes?: string; external_permit_number?: string },
): Promise<PermitRequest> {
  const response = await fetch(`${getApiBase()}/api/v1/agents/permit/requests/${id}/approve`, {
    method: "POST",
    headers: authHeaders(),
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new Error(await readError(response, "Ошибка утверждения"));
  return response.json();
}

export async function checkRosdorPermit(payload: {
  permit_number: string;
  plate: string;
}): Promise<{
  found: boolean;
  permit_number: string;
  plate: string;
  status: string | null;
  valid_until: string | null;
  route_summary: string | null;
  source: string;
}> {
  const response = await fetch(`${getApiBase()}/api/v1/integrations/rosdor/permits/check`, {
    method: "POST",
    headers: authHeaders(),
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new Error(await readError(response, "Ошибка проверки реестра"));
  return response.json();
}

export async function fetchRosdorHealth(): Promise<{
  status: string;
  message: string;
  configured: boolean;
}> {
  const response = await fetch(`${getApiBase()}/api/v1/integrations/rosdor/health`);
  if (!response.ok) throw new Error("Rosdor health failed");
  return response.json();
}

File: apps/api/app/schemas/permit.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.enums import PermitRequestStatus, PermitRiskLevel


class RoutePoint(BaseModel):
    name: str | None = None
    address: str | None = None
    lat: float | None = None
    lon: float | None = None


class CargoSpec(BaseModel):
    length_m: float = Field(gt=0, description="Длина груза, м")
    width_m: float = Field(gt=0, description="Ширина груза, м")
    height_m: float = Field(gt=0, description="Высота груза, м")
    mass_kg: float = Field(gt=0, description="Масса груза, кг")
    axle_loads_kg: list[float] = Field(default_factory=list, description="Нагрузка на оси, кг")


class PermitAnalysisResult(BaseModel):
    permit_required: bool
    podd_required: bool
    risk_level: PermitRiskLevel
    risk_score: float = Field(ge=0, le=1)
    route_summary: str
    restrictions: list[str] = Field(default_factory=list)
    special_conditions: list[str] = Field(default_factory=list)
    alternative_routes: list[str] = Field(default_factory=list)
    legal_references: list[str] = Field(default_factory=list)
    dispatcher_actions: list[str] = Field(default_factory=list)
    exceeds_limits: dict[str, bool] = Field(default_factory=dict)
    confidence: float = Field(ge=0, le=1, default=0.9)


class PermitRequestCreate(BaseModel):
    origin: RoutePoint
    destination: RoutePoint
    waypoints: list[RoutePoint] = Field(default_factory=list)
    cargo: CargoSpec
    vehicle_id: uuid.UUID | None = None
    order_id: uuid.UUID | None = None
    dispatcher_notes: str | None = None


class AnalyzeRouteRequest(BaseModel):
    origin: RoutePoint
    destination: RoutePoint
    waypoints: list[RoutePoint] = Field(default_factory=list)
    cargo: CargoSpec
    vehicle_id: uuid.UUID | None = None
    permit_request_id: uuid.UUID | None = None
    persist: bool = False


class AnalyzeRouteResponse(BaseModel):
    analysis: PermitAnalysisResult
    permit_request_id: uuid.UUID | None = None


class PermitApproveRequest(BaseModel):
    notes: str | None = None
    external_permit_number: str | None = None


class PermitRequestResponse(BaseModel):
    id: uuid.UUID
    organization_id: uuid.UUID
    order_id: uuid.UUID | None
    vehicle_id: uuid.UUID | None
    origin: RoutePoint
    destination: RoutePoint
    waypoints: list[RoutePoint]
    cargo: CargoSpec
    status: PermitRequestStatus
    analysis: PermitAnalysisResult | None
    risk_score: float | None
    external_permit_number: str | None
    dispatcher_notes: str | None
    approved_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PermitRequestListResponse(BaseModel):
    items: list[PermitRequestResponse]
    total: int


class RosdorHealthResponse(BaseModel):
    provider: str = "rosdor_monitoring"
    configured: bool
    parser_api_enabled: bool
    lk_enabled: bool
    status: str
    message: str


class RosdorPermitCheckRequest(BaseModel):
    permit_number: str = Field(min_length=3, max_length=64)
    plate: str = Field(min_length=3, max_length=16)


class RosdorPermitCheckResponse(BaseModel):
    found: bool
    permit_number: str
    plate: str
    status: str | None = None
    valid_until: str | None = None
    route_summary: str | None = None
    special_conditions: list[str] = Field(default_factory=list)
    source: str
